pdftext/pdf/chars.py

```diff
@@ -1,116 +1,52 @@
 import math
-from typing import List
 
+import pypdfium2 as pdfium
 import pypdfium2.raw as pdfium_c
-from pypdfium2 import PdfiumError
 
-from pdftext.pdf.utils import get_fontname, pdfium_page_bbox_to_device_bbox, page_bbox_to_device_bbox
-from pdftext.settings import settings
+from pdftext.pdf.utils import get_fontname
+from pdftext.schema import Bbox, Chars
 
 
-def update_previous_fonts(char_infos: List, i: int, prev_fontname: str, prev_fontflags: int, text_page, fontname_sample_freq: int):
-    min_update = max(0, i - fontname_sample_freq)  # Minimum index to update
-    for j in range(i - 1, min_update, -1):  # Goes from i to min_update
-        fontname, fontflags = get_fontname(text_page, j)
+def get_chars(textpage: pdfium.PdfTextPage, page_bbox: list[float], page_rotation: int, quote_loosebox=True, normalize=True) -> Chars:
+    chars: Chars = []
 
-        # If we hit the region with the previous fontname, we can bail out
-        if fontname == prev_fontname and fontflags == prev_fontflags:
-            break
-        char_infos[j]["font"]["name"] = fontname
-        char_infos[j]["font"]["flags"] = fontflags
+    x_start, y_start, x_end, y_end = page_bbox
+    page_width = math.ceil(abs(x_end - x_start))
+    page_height = math.ceil(abs(y_end - y_start))
 
+    for i in range(textpage.count_chars()):
+        fontname, fontflag = get_fontname(textpage, i)
+        text = chr(pdfium_c.FPDFText_GetUnicode(textpage, i))
 
-def flatten(page, flag=pdfium_c.FLAT_NORMALDISPLAY):
-    rc = pdfium_c.FPDFPage_Flatten(page, flag)
-    if rc == pdfium_c.FLATTEN_FAIL:
-        raise PdfiumError("Failed to flatten annotations / form fields.")
+        rotation = pdfium_c.FPDFText_GetCharAngle(textpage, i)
+        loosebox = rotation == 0 and (not text == "'" or quote_loosebox)
```

Comment:
Are there any other chars we need to special case?

Reply:
I tried a blanket special casing for control characters, other quote characters (unicodedata) but that only seems to be leading to worse results.